docs/troubleshooting.md

```diff
@@ -101,3 +102,44 @@ Don’t use distinct() as a metric with a large amount of data if you are using
+
+When saving/bookmarking a query consider the time range being used. When you come back to that query it will auto-run, and you will need to wait for it to complete to adjust the time range.
+
+### Limit your use of RegEx
+Using a regular expression (regex) allows for more flexible filtering, but be aware that these queries are more costly than if you filtered on the exact terms using `field IN (terms)`.
+
+There may be alternative fields that can be used to avoid the regex. E.g. tokenized field m instead of myTerm or a single URL segment instead of the full path.
+
+Frequently using regex to filter on a field may indicate a need for indexing more specialized views of that data in the index builder.
+
+Try to combine multiple regex filters on the same field into a single regex. E.g. `ref!=~"(apple|ifa_|orange123).*"` instead of `ref!=~"apple.*" ref!=~"ifa_.*" ref!=~"orange123.*"`
+
+RegEx filtering becomes particularly expensive when used on fields with a large number of distinct terms (e.g. myTerm), when terms are long (e.g. URLs) and when querying long time periods containing a large number of shards.
+
+### Don't try to un-invert an index / be smart about grouping
+IQL indices are stored inverted which makes it unfeasible to see the original documents as rows. Many users are used to tabular data and will try to un-invert an index by grouping by every available field. This is **not** recommended as it's very inefficient. If you need access to complete rows of data you should consider alternative data sources such as MySQL, HBase and LogRepo.
+
+If you think a grouping may result in a huge number of groups, you can get the actual number of expected groups by running a **DISTINCT** query: 
+
+`FROM example 1h today SELECT DISTINCT(stringField)` returns the number of groups you would get by running `FROM example 1h today GROUP BY stringField`
+
+Try to put the largest grouping as the last grouping. This allows the result of the last largest grouping to be streamed instead of all stored in memory. E.g., since there are dozens of somethings but millions of unique queries, prefer **`GROUP BY something, q`** to **`GROUP BY q, something`** or **~~`GROUP BY q[500000], something[50]`~~**.
+
+### Avoid distinct() when possible
+Avoid distinct() when possible for heavy queries. The optimization of streaming the data in the last group by grouping doesn't work if you have any distinct() elements in SELECT. So `FROM example 1h today select DISTINCT(stringField)` is fine but `FROM example 4w today GROUP BY q SELECT DISTINCT(stringField)` is bad.
+
+Although the output of DISTINCT(field) is a single number, all the values for that field have to be streamed from Imhotep to IQL, which can get really large over larger time ranges for fields with a lot of unique terms (e.g. myStringField, myExample, etc.)
+
+This also applies to IQL2 functions `SUM_OVER()`, `AVG_OVER()` and `DISTINCT_WINDOW()`.
```

Comment:
IQL2 is not available in OSS version yet. Remove this line.

Reply:
I handled this in #36 .